lib/Backends/CPU/Transforms.cpp

```diff
@@ -79,29 +79,56 @@ static Node *optimizeCPUConv(ConvolutionNode *CN, Function *F) {
       CN->getBias(), CN->getKernel(), CN->getStride(), CN->getPads(), group));
 }
 
+// Merge Max and Splat nodes into target-specific CPUMaxSplat node.
+// For quantized network, sinkRescaleQuantizedNode transformation might have
+// merged Rescale into Max node. In this case we need to pull it out, since
+// CPUMaxSplat requires input and output to be quantized the same way.
```

Comment:
I don't think this is the right fix, unless I miss something.
Indeed,  as far as I remember, the rescale node is not going to guarantee that its output will have the same offset and scale than the Max node.
E.g., consider
res = max(lhs, rhs)
It is possible to derive different ranges for lhs, rhs and res. In particular, res is likely to have to cover a smaller range than lhs and rhs.
In particular, there is nothing that ties the parameters of the quantization of rescale(lhs) to the parameters of the quantization of res.
This is all theoretical, in the sense that I am not versed into the gory details of our implementation of rescale and quantization in general, but just from the API (and checks in the verifier) it is possible to write something with different scale and offset for each operand (to be fair, we can also put a quantized and non-quantized operand, but that would crash somewhere :P). From the IR perspective, I believe we want this flexibility and the interpreter does the right thing as far as I can tell (dequantize individually each operand and re-quantize the result independently).

In fact, this issue highlights a bigger problem in Glow: what do we do with the legality of each operator?
Right now, each backend has to manually lower the IR to something that works for them (hence the proposed fix, though it wouldn't work I believe), however, it looks like we will want to describe such constraints in a high level way so that they can be fixed by a generic pass.

For instance, I wouldn't be surprised if a lot of backends require that both LHS and RHS have the same quantization parameters for all ALU, right now we cannot express that AFAICT.

@rdzhabarov Am I missing something or is it true that we cannot express something like `a = b + c` must have the same quantization parameters for a, b and c.

Reply:
"Right now, each backend has to manually lower the IR to something that works for them (hence the proposed fix, though it wouldn't work I believe), however, it looks like we will want to describe such constraints in a high-level way so that they can be fixed by a generic pass."
Interpreter, for instance, does not require any specialized handling of the max operator (quantized).
I do not have a good insight into other specific backends but could imagine that there might be ways to set {s,o} per LHS/RHS.

"Am I missing something or is it true that we cannot express something like a = b + c must have the same quantization parameters for a, b and c."
There is no direct way to do this, but in most of the cases, as you pointed out, quantization params would be different for a,b,c.
This specific CPU optimization is applied in order to avoid extra memory allocations for Splat but may introduce additional rescales (some rescales may cancel out with other rescales though). In general, it's not super clear if it's always worth it to enforce this behavior.

Reply:
> This specific CPU optimization is applied in order to avoid extra memory allocations for Splat but may introduce additional rescales (some rescales may cancel out with other rescales though). In general, it's not super clear if it's always worth it to enforce this behavior.

I agree and I thought it was even worse than that, I missed that the rescale node takes the need offset and scale that we want.

If that's just an optimization, then this patch is fine.

Reply:
You have a good point here. At the moment CPUMaxSplat requires input and output to be quantized the same way. It is so at the moment, but in the future this may change (as you described, the output of max is likely to have a smaller range and may be quantized differently in order to achieve better accuracy). So putting an assert here which checks that input type is equal to output type before creating CPUMaxSplat node maybe a good idea, to prevent people from debugging issues with it when/if quantization changes.